code/game/objects/items/devices/scanners/gas_analyzer.dm

```diff
@@ -0,0 +1,269 @@
+			return
+		to_chat(user, "<span class='notice'>The next [ongoing_weather] will hit in [butchertime(ongoing_weather.next_hit_time - world.time)].</span>")
+		if(ongoing_weather.aesthetic)
+			to_chat(user, "<span class='warning'>[src]'s barometer function says that the next storm will breeze on by.</span>")
+	else
+		var/next_hit = SSweather.next_hit_by_zlevel["[T.z]"]
+		var/fixed = next_hit ? next_hit - world.time : -1
+		if(fixed < 0)
+			to_chat(user, "<span class='warning'>[src]'s barometer function was unable to trace any weather patterns.</span>")
+		else
+			to_chat(user, "<span class='warning'>[src]'s barometer function says a storm will land in approximately [butchertime(fixed)].</span>")
+	cooldown = TRUE
+	addtimer(CALLBACK(src,/obj/item/analyzer/proc/ping), cooldown_time)
+
+/obj/item/analyzer/proc/ping()
+	if(isliving(loc))
+		var/mob/living/L = loc
+		to_chat(L, "<span class='notice'>[src]'s barometer function is ready!</span>")
+	playsound(src, 'sound/machines/click.ogg', 100)
+	cooldown = FALSE
+
+/// Applies the barometer inaccuracy to the gas reading.
+/obj/item/analyzer/proc/butchertime(amount)
+	if(!amount)
+		return
+	if(accuracy)
+		var/inaccurate = round(accuracy * (1 / 3))
+		if(prob(50))
+			amount -= inaccurate
+		if(prob(50))
+			amount += inaccurate
+	return DisplayTimeText(max(1, amount))
+
+/obj/item/analyzer/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "GasAnalyzer",  name, 500, 500, master_ui, state)
+		ui.open()
+
+/obj/item/analyzer/ui_data(mob/user)
+	var/list/data = list()
+	if(auto_updating)
+		on_analyze(source=src, target=scan_target)
+	LAZYINITLIST(last_gasmix_data)
+	LAZYINITLIST(history_gasmix_data)
+	data["gasmixes"] = last_gasmix_data
+	data["autoUpdating"] = auto_updating
+	data["historyGasmixes"] = history_gasmix_data
+	data["historyViewMode"] = history_view_mode
+	data["historyIndex"] = history_gasmix_index
+	return data
+
+/obj/item/analyzer/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+	switch(action)
+		if("autoscantoggle")
+			auto_updating = !auto_updating
+			return TRUE
+		if("input")
+			if(!length(history_gasmix_data))
+				return TRUE
+			var/target = text2num(params["target"])
+			auto_updating = FALSE
+			history_gasmix_index = target
+			last_gasmix_data = history_gasmix_data[history_gasmix_index]
+			return TRUE
+		if("clearhistory")
+			history_gasmix_data = list()
+			return TRUE
+		if("modekpa")
+			history_view_mode = ANALYZER_HISTORY_MODE_KPA
+			return TRUE
+		if("modemol")
+			history_view_mode = ANALYZER_HISTORY_MODE_MOL
+			return TRUE
+
+/// Called when our analyzer is used on something
+/obj/item/analyzer/proc/on_analyze(datum/source, atom/target, save_data=TRUE)
+	SIGNAL_HANDLER
```

Comment:
This isn't called via a signal. Please clean up the arg from this and from all calls of the proc.
```suggestion
/obj/item/analyzer/proc/on_analyze(atom/target, save_data = TRUE)
```